pandas/core/apply.py

```diff
@@ -158,19 +159,54 @@ def agg(self) -> DataFrame | Series | None:
             return self.apply_str()
 
         if is_dict_like(arg):
-            return self.agg_dict_like()
+            return self.dict_like("agg")
         elif is_list_like(arg):
             # we require a list, but not a 'str'
-            return self.agg_list_like()
+            return self.list_like("agg")
 
         if callable(arg):
             f = com.get_cython_func(arg)
             if f and not args and not kwargs:
                 return getattr(obj, f)()
+            elif not isinstance(obj, SelectionMixin):
+                # i.e. obj is Series or DataFrame
+                return self.agg_udf()
 
         # caller can react
         return None
 
+    def agg_udf(self):
+        obj = self.obj
+        arg = cast(Callable, self.f)
+
+        if not isinstance(obj, SelectionMixin):
+            # i.e. obj is Series or DataFrame
+            selected_obj = obj
+        elif obj._selected_obj.ndim == 1:
+            # For SeriesGroupBy this matches _obj_with_exclusions
+            selected_obj = obj._selected_obj
+        else:
+            selected_obj = obj._obj_with_exclusions
+
+        results = []
+
+        if selected_obj.ndim == 1:
+            colg = obj._gotitem(selected_obj.name, ndim=1, subset=selected_obj)
+            return arg(colg)
+
+        indices = []
+        for index, col in enumerate(selected_obj):
+            colg = obj._gotitem(col, ndim=1, subset=selected_obj.iloc[:, index])
+            new_res = arg(colg)
+            results.append(new_res)
+            indices.append(index)
+        keys = selected_obj.columns.take(indices)
+
+        from pandas import Series
+
+        result = Series(results, index=keys)
```

Comment:
should we be using something._constructor here?

Reply:
We could be using `selected_obj._constructor_sliced`